io/io.go

```diff
@@ -0,0 +1,327 @@
+	ReadDir(n int) ([]fs.DirEntry, error)
+}
+
+// FS wraps an io/fs.FS as an IO interface.
+func FS(fsys fs.FS) IO {
+	if _, ok := fsys.(fs.ReadFileFS); ok {
+		return readFileFS{ioFS{fsys, nil}}
+	}
+	return ioFS{fsys, nil}
+}
+
+// FSPreProcName wraps an io/fs.FS like FS, only if fn is non-nil then
+// it is called to preprocess any filenames before they are passed to
+// the underlying fsys.
+func FSPreProcName(fsys fs.FS, fn func(string) string) IO {
+	if _, ok := fsys.(fs.ReadFileFS); ok {
+		return readFileFS{ioFS{fsys, fn}}
+	}
+	return ioFS{fsys, fn}
+}
+
+type readFileFS struct {
+	ioFS
+}
+
+func (r readFileFS) ReadFile(name string) ([]byte, error) {
+	if r.preProcessName != nil {
+		name = r.preProcessName(name)
+	}
+
+	rfs, ok := r.fsys.(fs.ReadFileFS)
+	if !ok {
+		return nil, errMissingReadFile
+	}
+	return rfs.ReadFile(name)
+}
+
+type ioFS struct {
+	fsys fs.FS
+
+	preProcessName func(string) string
+}
+
+func (f ioFS) Open(name string) (File, error) {
+	if f.preProcessName != nil {
+		name = f.preProcessName(name)
+	}
+
+	if name == "/" {
+		name = "."
+	} else {
+		name = strings.TrimPrefix(name, "/")
+	}
+	file, err := f.fsys.Open(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return ioFile{file}, nil
+}
+
+func (f ioFS) Remove(name string) error {
+	r, ok := f.fsys.(interface{ Remove(name string) error })
+	if !ok {
+		return errMissingRemove
+	}
+	return r.Remove(name)
+}
+
+var (
+	errMissingReadDir  = errors.New("fs.File directory missing ReadDir method")
+	errMissingSeek     = errors.New("fs.File missing Seek method")
+	errMissingReadAt   = errors.New("fs.File missing ReadAt")
+	errMissingRemove   = errors.New("fs.FS missing Remove method")
+	errMissingReadFile = errors.New("fs.FS missing ReadFile method")
+)
+
+type ioFile struct {
+	file fs.File
+}
+
+func (f ioFile) Close() error               { return f.file.Close() }
+func (f ioFile) Read(b []byte) (int, error) { return f.file.Read(b) }
+func (f ioFile) Stat() (fs.FileInfo, error) { return f.file.Stat() }
+func (f ioFile) Seek(offset int64, whence int) (int64, error) {
+	s, ok := f.file.(io.Seeker)
+	if !ok {
+		return 0, errMissingSeek
+	}
+	return s.Seek(offset, whence)
+}
+
+func (f ioFile) ReadAt(p []byte, off int64) (n int, err error) {
+	r, ok := f.file.(io.ReaderAt)
+	if !ok {
+		return 0, errMissingReadAt
+	}
+	return r.ReadAt(p, off)
+}
+
+func (f ioFile) ReadDir(count int) ([]fs.DirEntry, error) {
+	d, ok := f.file.(fs.ReadDirFile)
+	if !ok {
+		return nil, errMissingReadDir
+	}
+
+	return d.ReadDir(count)
+}
+
+func inferFileIOFromSchema(path string, props map[string]string) (IO, error) {
+	parsed, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	switch parsed.Scheme {
+	case "s3", "s3a", "s3n":
+		opts := []func(*config.LoadOptions) error{}
+		endpoint, ok := props[S3EndpointURL]
+		if !ok {
+			endpoint = os.Getenv("AWS_S3_ENDPOINT")
+		}
+
+		if endpoint != "" {
```

Comment:
I would rather move all of this s3-related things to s3.go or something. Otherwise this code here will get quite difficult to read later when e.g. gcs or azure support will be added

Reply:
That's a fair point. I'll update this shortly